torchtune/datasets/rfc_iterable_dataset.md

```diff
@@ -0,0 +1,439 @@
+# Alternative: use dataclasses?
+
+def setup_data(
+    dataset_cfg: ConfigDict,
+    dataset_defaults: ConfigDict,
+    data_setup_cfg: ConfigDict,
+    dataloader_cfg: ConfigDict,
+    seed: int,
+    pad_idx: int,
+    ignore_idx: int,
+    pad_to_multiple_of: int,
+) -> "IterableDataset":
+    """Equivalent to setup_data in the recipe."""
+    iterable_datasets = []
+    weights = []
+    dataset_defaults = {} if dataset_defaults is None else dataset_defaults
+
+    # Add dataset to a list just for processing
+    if not isinstance(dataset_cfg, list):
+        dataset_cfg = [dataset_cfg]
+
+    # instantiate
+    for base_cfg in dataset_cfg:
+        weight = base_cfg.get("weight", 1.0)
+        weights.append(weight)
+
+        base_cfg = OmegaConf.merge(dataset_defaults, base_cfg)
+        ds = instantiate(base_cfg)
+        iterable_datasets.append(ds)
+
+    # Interleave for multidataset
+    if len(iterable_datasets) > 1:
+        weights = normalize_weights(weights)  # sum to 1
+        ds = interleave_datasets(
+            iterable_datasets,
+            probabilities=weights,
+            seed=seed,
+            # strategies: https://huggingface.co/docs/datasets/v3.3.2/en/package_reference/main_classes#datasets.interleave_datasets.stopping_strategy
+            stopping_strategy=data_setup_cfg.multidataset_stopping_strategy,
+        )
+    else:
+        ds = iterable_datasets[0]
+
+    if setup_cfg.packing:
+        # Subclass of IterableDataset, takes any iterator as input
+        ds = instantiate(
+            data_setup_cfg.packing,
+            dataset=ds,
+            padding_idx=pad_id,  # TODO: in the future, move padding to collate_fn
+        )
+
+    # Instantiate collate_fn
+    collate_fn = dataloader_cfg.pop("collate_fn", None)
+    # TODO: in the future, unify those two
+    if collate_fn is None:
+        collate_fn = (
+            "torchtune.data.padded_collate_packed"
+            if packing else
+            "torchtune.data.padded_collate_sft"
+        )
+
+    collate_fn = _get_component_from_path(collate_fn)
+    collate_fn = partial(
+        collate_fn,
+        padding_idx=pad_idx,
+        ignore_idx=ignore_id,
+        pad_to_multiple_of=pad_to_multiple_of
+    )
+
+    # Dropping last avoids shape issues with compile + flex attention
+    if "drop_last" not in dataloader_cfg:
+        dataloader_cfg["drop_last"] = True
+
+    dataloader = instantiate(dataloader_cfg, dataset=ds, collate_fn=collate_fn)
+
+    return dataloader
+```
+
+#### Recipe Train Loop
+
+```python
+for epoch in range(n_epochs):
+    # TODO: review the proper way to reshuffle after each epoch
+    # https://huggingface.co/docs/datasets/v3.6.0/en/about_mapstyle_vs_iterable#speed-differences
+    self._dataloader.set_epoch(epoch)
+    for example in dataloader:
+        pass
+```
+
+### Backward Compatibility
+
+Options:
+
+1. Make setup_data an utility, and have two utilities supporting old and new config formats.
```

Comment:
Is there idea here to have something like `recipes/recipe_utils.py` that would contain all these builder functions?

Or maybe it's put in `torchtune` core? 

I'm not sure how to handle this cc @pbontrager